### ISSUE TITLE
How to show errors when building fails?

### ISSUE THREAD

**Issue by TellNoLies** [TeXitoi/ortho60-keyberon#11]
Hello,
I am trying to make another layout without the last one that I had made.
When in the src/ folder with the layout.rs I use the build command, "cargo objcopy --bin ortho60-keyberon --release -- -O binary ortho60-keyberon.bin" I get this error: "Compiling ortho60-keyberon v0.1.0 (/home/rbear/Build/ortho60-keyberon)
error: could not compile `ortho60-keyberon` due to 12 previous errors"
However I have no clue what I messed up on. I am guessing it is a keycode that I got wrong but not sure on which ones. It builds the default layout.rs fine, so I know that I installed the dependencies properly.
My changes:

```rust
use keyberon::action::Action::Trans;
use keyberon::action::{k, l, m};
use keyberon::key_code::KeyCode::*;

// Shift + KeyCode
macro_rules! s {
    ($k:ident) => {
        m(&[LShift, $k])
    };
}

#[rustfmt::skip]
pub static LAYERS: keyberon::layout::Layers = &[
    &[
        &[k(Grave), k(Kb1),k(Kb2),k(Kb3), k(Kb4), k(Kb5),  k(Kb6),  k(Kb7),k(Kb8), k(Kb9),  k(Kb0),   k(Pipe)],
        &[k(Tab),   k(Q),  k(W),  k(F),   k(P),   k(B),    k(J),    k(L),  k(U),   k(Y),    k(SColon),k(BSlash)],
        &[k(BSpace),k(A),  k(R),  k(S),   k(T),   k(G),    k(M),    k(N),  k(E),   k(I),    k(O),     k(Quote) ],
        &[k(LShift),k(Z),  k(X),  k(C),   k(D),   k(V),    k(K),    k(H),  k(Comma),k(Dot), k(Slash), k(Enter) ],
        &[k(LCtrl), k(LAlt),k(LGui),k(LShift),l(1),k(Space),k(Space),l(2), k(RCtrl), k(RShift), k(Escape), k(RGui) ],
    ], &[
        &[s!(Grave), s!(Kb1),   s!(Kb2),   s!(Kb3),     s!(Kb4),      s!(Kb5),   s!(Kb6), s!(Kb7), s!(Kb8),s!(Kb9), s!(Kb0),     k(PScreen)      ],
        &[k(Tab),    k(PgUp),   k(End),     k(Up),       k(Home),      k(Insert), Trans,   k(Kb7), k(Kb8), k(Kb9),  k(Kb0),      k(CapsLock)     ],
        &[k(Delete), k(PgDown), k(Left),    k(Down),     k(Right),     k(Select),  Trans,   k(Kb4), k(Kb5), k(Kb6),  Trans, k(Bslash)       ],
        &[Trans,     k(ScrollLock), Trans, m(&[LAlt, Left]), m(&[LAlt, Right]),Trans,k(Kb0), k(Kb1), k(KB2), k(KB3), k(MediaMute),   Trans       ],
        &[Trans,     Trans,     Trans,      Trans,       Trans,  Trans,  Trans, k(MediaPreviousSong),k(MediaNextSong),k(VolDown),k(VolUp),k(MediaPlayPause)],
    ], &[
        &[k(Escape), k(Kb1), k(Kb2), k(Kb3), k(Kb4),  k(Kb5),  k(Kb6),    k(Kb7),  k(Kb8),   k(Kb9),     k(Kb0),    k(Pipe)        ],
        &[s!(Grave), s!(Kb1),s!(Kb2),s!(Kb3), s!(Kb4), s!(Kb5), s!(Kb6),  s!(Kb7),  s!(Kb8), s!(Kb9),    s!(Kb0),   k(CapsLock)    ],
        &[k(F1),     k(F2),  k(F3),  k(F4),  k(F5),   k(F6),  s!(Kb9),   s!(Kb0),  s!(Minus), k(Minus), k(Plus), k(Equal)     ],
        &[k(F7),     k(F8),  k(F9),  k(F10), k(F11),  k(F12), k(LBrace), k(RBrace),s!(LBrace), s!(RBrace), s!(Comma), s!(Dot)         ],
        &[Trans,     Trans,  Trans,  Trans,  Trans,   Trans,   l(3),      Trans,   m(&[RCtrl, RAlt]),  Trans,  Trans,   Trans         ],
    ], &[
        &[k(F1), k(F2),   k(F3),   k(F4),  k(F5),  k(F6),  k(F7),   k(F8),   k(F9),   k(F10),  k(F11),   k(F12)        ],
        &[Trans, Trans,  Trans,  Trans, Trans, Trans,                                          k(NumLock), k(Kp7), k(Kp8), k(Kp9), k(KpPlus), k(KpSlash)        ],
        &[k(Mute),k(MediaPreviousSong),k(VolDown),k(Volup),k(MediaNextSong),k(MediaPlayPause), k(KpComma), k(Kp4), k(Kp5), k(Kp6), k(KpMinus),k(KpAsterisk)     ],
        &[k(Find),k(Stop), k(Again), k(Undo),k(Cut),k(Paste),                                     k(Kp0),  k(Kp1), k(Kp2), k(Kp3), k(KpDot),  k(KpEnter)        ],
        &[Trans,  Trans, Trans, Trans,  Trans,  Trans,  l(3),  l(2),   Trans,   Trans,  Trans,    Trans  ],
    ]

]; 
```

Thanks for the help.

**Comment by TeXitoi** [comment]
`cargo build --release` will show the errors.

**Comment by TeXitoi** [comment]
```
   Compiling ortho60-keyberon v0.1.0 (/home/gpinot/dev/ortho60-keyberon)
error[E0425]: cannot find value `Pipe` in this scope
  --> src/layout.rs:15:105
   |
15 |         &[k(Grave), k(Kb1),k(Kb2),k(Kb3), k(Kb4), k(Kb5),  k(Kb6),  k(Kb7),k(Kb8), k(Kb9),  k(Kb0),   k(Pipe)],
   |                                                                                                         ^^^^ not found in this scope

error[E0425]: cannot find value `BSlash` in this scope
  --> src/layout.rs:16:105
   |
16 |         &[k(Tab),   k(Q),  k(W),  k(F),   k(P),   k(B),    k(J),    k(L),  k(U),   k(Y),    k(SColon),k(BSlash)],
   |                                                                                                         ^^^^^^ help: a unit variant with a similar name exists (notice the capitalization): `Bslash`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:78:5
   |
78 |     Bslash,
   |     ------ similarly named unit variant `Bslash` defined here

error[E0425]: cannot find value `KB2` in this scope
  --> src/layout.rs:24:104
   |
24 | ...LAlt, Right]),Trans,k(Kb0), k(Kb1), k(KB2), k(KB3), k(MediaMute),   Trans       ],
   |                                          ^^^ help: a unit variant with a similar name exists: `Kb2`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:47:5
   |
47 |     Kb2,
   |     --- similarly named unit variant `Kb2` defined here

error[E0425]: cannot find value `KB3` in this scope
  --> src/layout.rs:24:112
   |
24 | ...ght]),Trans,k(Kb0), k(Kb1), k(KB2), k(KB3), k(MediaMute),   Trans       ],
   |                                          ^^^ help: a unit variant with a similar name exists: `Kb3`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:49:5
   |
49 |     Kb3, // 0x20
   |     --- similarly named unit variant `Kb3` defined here

error[E0425]: cannot find value `Pipe` in this scope
  --> src/layout.rs:27:119
   |
27 |         &[k(Escape), k(Kb1), k(Kb2), k(Kb3), k(Kb4),  k(Kb5),  k(Kb6),    k(Kb7),  k(Kb8),   k(Kb9),     k(Kb0),    k(Pipe)        ],
   |                                                                                                                       ^^^^ not found in this scope

error[E0425]: cannot find value `Plus` in this scope
  --> src/layout.rs:29:107
   |
29 |         &[k(F1),     k(F2),  k(F3),  k(F4),  k(F5),   k(F6),  s!(Kb9),   s!(Kb0),  s!(Minus), k(Minus), k(Plus), k(Equal)     ],
   |                                                                                                           ^^^^ not found in this scope

error[E0425]: cannot find value `LBrace` in this scope
  --> src/layout.rs:30:65
   |
30 |         &[k(F7),     k(F8),  k(F9),  k(F10), k(F11),  k(F12), k(LBrace), k(RBrace),s!(LBrace), s!(RBrace), s!(Comma), s!(Dot)         ],
   |                                                                 ^^^^^^ help: a unit variant with a similar name exists: `LBracket`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:74:5
   |
74 |     LBracket,
   |     -------- similarly named unit variant `LBracket` defined here

error[E0425]: cannot find value `RBrace` in this scope
  --> src/layout.rs:30:76
   |
30 |         &[k(F7),     k(F8),  k(F9),  k(F10), k(F11),  k(F12), k(LBrace), k(RBrace),s!(LBrace), s!(RBrace), s!(Comma), s!(Dot)         ],
   |                                                                            ^^^^^^ help: a unit variant with a similar name exists: `RBracket`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:76:5
   |
76 |     RBracket, // 0x30
   |     -------- similarly named unit variant `RBracket` defined here

error[E0425]: cannot find value `LBrace` in this scope
  --> src/layout.rs:30:87
   |
30 |         &[k(F7),     k(F8),  k(F9),  k(F10), k(F11),  k(F12), k(LBrace), k(RBrace),s!(LBrace), s!(RBrace), s!(Comma), s!(Dot)         ],
   |                                                                                       ^^^^^^ help: a unit variant with a similar name exists: `LBracket`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:74:5
   |
74 |     LBracket,
   |     -------- similarly named unit variant `LBracket` defined here

error[E0425]: cannot find value `RBrace` in this scope
  --> src/layout.rs:30:99
   |
30 |         &[k(F7),     k(F8),  k(F9),  k(F10), k(F11),  k(F12), k(LBrace), k(RBrace),s!(LBrace), s!(RBrace), s!(Comma), s!(Dot)         ],
   |                                                                                                   ^^^^^^ help: a unit variant with a similar name exists: `RBracket`
   | 
  ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:76:5
   |
76 |     RBracket, // 0x30
   |     -------- similarly named unit variant `RBracket` defined here

error[E0425]: cannot find value `Volup` in this scope
   --> src/layout.rs:35:53
    |
35  | ...ousSong),k(VolDown),k(Volup),k(MediaNextSong),k(MediaPlayPause), k(KpComma), k(Kp4), k(Kp5), k(Kp6), k(KpMinus),k(KpAsterisk)     ],
    |                          ^^^^^ help: a unit variant with a similar name exists (notice the capitalization): `VolUp`
    | 
   ::: /home/gpinot/.cargo/registry/src/github.com-1ecc6299db9ec823/keyberon-0.1.0/src/key_code.rs:174:5
    |
174 |     VolUp, // 0x80
    |     ----- similarly named unit variant `VolUp` defined here

For more information about this error, try `rustc --explain E0425`.
error: could not compile `ortho60-keyberon` due to 11 previous errors
```

**Comment by TellNoLies** [comment]
Thank you very much.
Is this the correct place to get the keycodes?
[https://docs.rs/keyberon/0.1.1/keyberon/key_code/enum.KeyCode.html](url)

**Comment by TeXitoi** [comment]
yup. don't forget #10

**Comment by TellNoLies** [comment]
That is what I just ran into with this error I assume.
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 1eaf:0003
Run-time device DFU version 0110
Claiming USB DFU Interface...
Setting Alternate Setting #2 ...
dfu-util: Cannot set alternate interface

Is it that I just need to have the most current rust build, even though I just built it yesterday?

**Comment by TellNoLies** [comment]
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Warning: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release
dfu-util: Failed to retrieve language identifiers
dfu-util: Failed to retrieve language identifiers
Opening DFU capable USB device...
Device ID 1eaf:0003
Device DFU version 0110
Claiming USB DFU Interface...
Setting Alternate Interface #2 ...
dfu-util: Cannot set alternate interface: LIBUSB_ERROR_OTHER

It changed with dfu-utill 11

**Comment by TeXitoi** [comment]
This is unrelated, dfu-util doesn't detect your bluepill. Do you have an already working ortho60? Because flashing the rust firmware is just the same as flashing the cannonkeys firmware.

You need to use an old version of rust for keyberon 0.1 to work, or you need to upgrade to keyberon master, as specified in #10 

And, to be sure, you have a ST-Link to be able to recover from buggy flash?

**Comment by TellNoLies** [comment]
Ah, Okay. 
It was working until I attempted flash my board. Now there's nothing but the red led on the pill.
I will dig out my st-link to see if I can get it to work again.

**Comment by TeXitoi** [comment]
looks like #10

**Comment by TeXitoi** [comment]
I think if you press reset, you might still reboot in DFU mode, but not sure.

**Comment by TellNoLies** [comment]
It has been a while since I have done this. Is this the command to reset the bluepill?
openocd -f interface/stlink-v2.cfg -f target/stm32f1x.cfg -c "init; reset halt; stm32f1x mass_erase 0; program generic_boot20_pc13.bin exit 0x08000000"

**Comment by TeXitoi** [comment]
Yes, it erases everything and flash the bootloader.

**Comment by TellNoLies** [comment]
st-info --probe
Found 1 stlink programmers
  version:    V2J17S4
  serial:     56FF70066580515215170667
  flash:      131072 (pagesize: 1024)
  sram:       20480
  chipid:     0x0410
  descr:      F1xx Medium-density
13.bin exit 0x08000000"ce/stlink-v2.cfg -f target/stm32f1x.cfg -c "init; reset halt; stm32f1x mass_erase 0; program generic_boot20_pc1
Open On-Chip Debugger 0.10.0
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
adapter speed: 1000 kHz
adapter_nsrst_delay: 100
none separate
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : clock speed 950 kHz
Info : STLINK v2 JTAG v17 API v2 SWIM v4 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 3.156573
Info : stm32f1x.cpu: hardware has 6 breakpoints, 4 watchpoints
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x080000f0 msp: 0x20005000
Info : device id = 0x20036410
Info : flash size = 128kbytes
stm32x mass erase complete
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0xfffffffe msp: 0xfffffffc
** Programming Started **
auto erase enabled
Error: couldn't open generic_boot20_pc13.bin
** Programming Failed **
shutdown command invoked


Is the error at the end okay or did it fail to clear the bootloader?

**Comment by TeXitoi** [comment]
you didn't flash the bootloader as it was not in the working directory.

**Comment by TellNoLies** [comment]
(edit) I change to the ortho60-keyberon and was able to re-flash the bootloader.
Okay I got it to flash but none of the keys are working.
This the output:
dfu-util -d 1eaf:0003 -a 2 -D ortho60-keyberon.bin
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Warning: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release
Opening DFU capable USB device...
Device ID 1eaf:0003
Device DFU version 0110
Claiming USB DFU Interface...
Setting Alternate Interface #2 ...
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 0110
Device returned transfer size 1024
Copying data from PC to DFU device
Download        [=========================] 100%        26836 bytes
Download done.
DFU state(8) = dfuMANIFEST-WAIT-RESET, status(0) = No error condition is present
Resetting USB to switch back to runtime mode
Done!

It seems to have flashed but no output.
(edit) I even flashed the default layout and there still is nothing.

**Comment by TeXitoi** [comment]
#10 downgrade your compiler.

**Comment by TeXitoi** [comment]
Easier fix: switch to rust 1.47

**Comment by TellNoLies** [comment]
When I rebuild rust, do I use the one before "beta"?
(edit) When I ran the stable it installs 1.56. 
How do I chose the version? I am running "curl https://sh.rustup.rs -sSf | sh" to install the version. Is this the correct way?
I have no clue on how to do it looking at the other posts. I do not see a command mentioned on how to select version.
Thank you for your help and patience.

**Comment by TeXitoi** [comment]
```
rustup toolchain install 1.47
cargo +1.47 ...
```

**Comment by TellNoLies** [comment]
Thank you. 
I get this error though.
rustup toolchain install 1.47
warning: downloading with complete profile isn't recommended unless you are a developer of the rust language
info: syncing channel updates for '1.47-x86_64-unknown-linux-gnu'
info: latest update on 2020-10-08, rust version 1.47.0 (18bf6b4f0 2020-10-07)
error: some components unavailable for download for channel '1.47': 'miri' for target 'x86_64-unknown-linux-gnu', 'rust-analyzer-preview' for target 'x86_64-unknown-linux-gnu'If you don't need the components, you could try a minimal installation with:

    rustup toolchain add 1.47 --profile minimal
ortho60-keyberon on  master [!?] via 🦀 v1.56.0   cargo +1.47 ...
error: toolchain '1.47-x86_64-unknown-linux-gnu' is not installed

**Comment by TellNoLies** [comment]
Well, still no dice.
dfu-util -d 1eaf:0003 -a 2 -D ortho60-keyberon.bin
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Warning: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release
Opening DFU capable USB device...
Device ID 1eaf:0003
Device DFU version 0110
Claiming USB DFU Interface...
Setting Alternate Interface #2 ...
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 0110
Device returned transfer size 1024
Copying data from PC to DFU device
Download        [=========================] 100%        26808 bytes
Download done.
DFU state(8) = dfuMANIFEST-WAIT-RESET, status(0) = No error condition is present
Resetting USB to switch back to runtime mode
Done!

I didn't get any errors but still no output from the keys.

**Comment by TellNoLies** [comment]
rustup toolchain install 1.47
cargo +1.47 ...
warning: downloading with complete profile isn't recommended unless you are a developer of the rust language
info: syncing channel updates for '1.47-x86_64-unknown-linux-gnu'

  1.47-x86_64-unknown-linux-gnu unchanged - rustc 1.47.0 (18bf6b4f0 2020-10-07)

info: checking for self-updates
error: no such subcommand: `...`

        Did you mean `b`?

**Comment by TellNoLies** [comment]
rustup -V
rustup 1.24.3 (ce5817a94 2021-05-31)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.56.0 (09c42c458 2021-10-18)

**Comment by TeXitoi** [comment]
... means whatever you use cargo, so `build --release`, `obj...`

you will also need to install arm components for your 1.47 as explained on the README. Simple would be `rustup default 1.47`, and then installing components as explained in the README, and then compiling as explained in the README

**Comment by TeXitoi** [comment]
you can then go back to stable with `rustup default stable`

**Comment by TellNoLies** [comment]
I am sorry that I kept bugging you on this.
What I did was removed the keyberon git folders and re-ran the installation process in the readme. Because, no mater how many times I ran the commands in the instructions it wouldn't work. After a fresh start though, I ran all the deps which said they where the latest versions. Finally, I rebuilt the custom layout, flashed it, and now it is working!!.

Thank you so much for your help.
